### PR TITLE
Implement pretty printing for dataset objects

### PR DESCRIPTION
**Pull request by pvandyken**
Just wanted to give people a chance to look at this, I still have to write appropriate tests. And there's some wonky edge cases that I have to fix.

Here's an example output:

```py
BidsDataset({
    "├ƒ": BidsComponent(
        name="├ƒ",
        path="mod-{mod}_echo-{echo}_space-{space}",
        zip_lists={
            "echo":  ["91278", "32",   "91278", ... "3",    "3",       "91278", "32",   "91278"],
            "space": ["YW3N",  "BlPR", "BlPR",  ... "BlPR", "YW3N",    "z7j54", "YW3N", "z7j54"],
            "mod":   ["c",     "Ck",   "Ck",    ... "Ck",   "jHde1Az", "c",     "c",    "Ck"   ],
        },
    ),
    "├çΣ╡ñY├╢≥╖¬Ç≤╢¡╕├╝": BidsComponent(
        name="├çΣ╡ñY├╢≥╖¬Ç≤╢¡╕├╝",
        path="echo-{echo}_space-{space}",
        zip_lists={
            "echo":  ["6388",     "04236"   ],
            "space": ["o6yjcrZa", "o6yjcrZa"],
        },
    ),
})
```

The string is formatted to be valid python code, but with newlines and alignment of the entity-values. In interactive environments only (not in scripts), the values will elide to fit the screen, as shown in the example above.

If you want to try for yourself, the easiest way is to start an REPL and call `snakebids.tests.strategies.datasets().example()`.

In principle, we could do pretty printing only when in an interactive environment, but as I can only really imagine this output being used for debugging, I think it's worth keeping it like this.

Let me know if you have any comments.

### CODE REVIEW

**Review by tkkuehn — COMMENTED** [review]
After a brief, non-exhaustive, look at the code and the output you posted, this looks good to me in principle.

This seems like it would have been a chore to write -- thanks for taking it on.

**Comment by pvandyken** [comment]
Currently I have the pretty-printing logic contained a method called `format()` (e.g. `BidsComponent.format()`). I'd like to change this, as that name is too valuable for just printing, and maybe we'll want it later. 

Alternatives:
* `format_print`
* `format_str`
* `dump` (I've seen this in other libraries)
* `info` (Probably not... an "info" view is different than a `__repr__`)
* `pretty`

I'm sure there's other possibilities. Thoughts? @akhanf, @tkkuehn, @kaitj

**Comment by kaitj** [comment]
What about borrowing from the `pprint` module and calling it `pformat()`?

**Comment by pvandyken** [comment]
I like that; made the change. This is ready for review

**Comment by pvandyken** [comment]
Compared to the initial example, the ellipses `...` will now be biased toward the end of the list. So in a list elided down to four elements plus ellipse, the ellipse will be in the centre (index 2), in a list elided to 5 elements plus ellipse, the ellipse will be one to the right of centre (index 3)

**Review by tkkuehn — APPROVED** [review]
One pretty minor comment that you can ignore if you prefer. Thanks again!

File: snakebids/core/datasets.py [hunk 1]
```diff
@@ -55,6 +59,23 @@ class BidsComponent:
         on_setattr=attr.setters.frozen, converter=dict
     )
 
+    def __repr__(self) -> str:
```
[the remark this answers is not in the document]
Comment (pvandyken) [reply snakebids/core/datasets.py:62]: Hmmm, so the challenge is if you call the object in the REPL by itself (e.g `>>> component`), it's `__repr__` that gets called, not `__str__`. So if we made this change, to take advantage of the width restriction, you would need to call `print(component)`. Since these methods were in large part designed to provide convenience for the REPL, I went ahead and applied width-restricted formatting to `__repr__`. Thoughts on this? (For what it's worth, I don't think numpy, pandas, or xarray follow this guidance from the python docs, as it would lead to enormously big outputs)

[the remark this answers is not in the document]
Comment (tkkuehn) [reply snakebids/core/datasets.py:62]: Ah that's a good enough reason to leave it as is in my opinion!
